Web3/Classes/Core/Transaction/EthereumTransaction.swift

```diff
@@ -9,7 +9,116 @@
 import Foundation
 import BigInt
 
-public struct EthereumTransaction {
+public struct EthereumTransaction: Codable {
+    /// The number of transactions made prior to this one
+    public var nonce: EthereumQuantity?
+    
+    /// Gas price provided Wei
+    public var gasPrice: EthereumQuantity?
+    
+    /// Gas limit provided
+    public var gasLimit: EthereumQuantity?
+    
+    /// Address of the sender
+    public var from: EthereumAddress?
+    
+    /// Address of the receiver
+    public var to: EthereumAddress?
+    
+    /// Value to transfer provided in Wei
+    public var value: EthereumQuantity?
+    
+    /// Input data for this transaction
+    public var data: EthereumData
+    
+    // MARK: - Initialization
+    
+    /**
+     * Initializes a new instance of `EthereumTransaction` with the given values.
+     *
+     * - parameter nonce: The nonce of this transaction.
+     * - parameter gasPrice: The gas price for this transaction in wei.
+     * - parameter gasLimit: The gas limit for this transaction.
+     * - parameter from: The address to send from, required to send a transaction using sendTransaction()
+     * - parameter to: The address of the receiver.
+     * - parameter value: The value to be sent by this transaction in wei.
+     * - parameter data: Input data for this transaction. Defaults to [].
+     */
+    init(
+        nonce: EthereumQuantity? = nil,
+        gasPrice: EthereumQuantity? = nil,
+        gasLimit: EthereumQuantity? = nil,
+        from: EthereumAddress? = nil,
```

Comment:
Should `from` be an Optional? What does the node do if you don't provide a `from`?

Reply:
So the challenge with this is that from is required for submitting to web3.sendTransaction but it is not good for people who are signing locally. I ran into this issue when I was updating the test for signing a transaction. A couple thoughts I had:

One solution would be to simply validate this in the Eth sendTransaction method.

Another could be that from is a parameter in the eth.sendTransaction method, but this would be a departure from the web3.js api. 

A third option would be to rethink this PR and make a struct for composing a transaction called EthereumTransactionAttributes where everything is optional, and perhaps deal with signed transactions another way.

I think it comes down to what we expect the purpose of EthereumTransaction to be. Do you have a preference for any of these solutions?

Reply:
It looks like the strategy for web3.js is that they use an inputFormatter on the object passed to automatically fill in the `from` field with the defaultAccount if one is not included already, and if there is no default account then it will throw an error: https://github.com/ethereum/web3.js/blob/1.0/packages/web3-core-helpers/src/formatters.js#L138

Reply:
What if we added a static function to `EthereumAddress` which returns a zero address. Like the following:

```Swift
extension EthereumAddress {

    static func zero() throws -> EthereumAddress {
        return try EthereumAddress(bytes: Bytes(repeating: 0, count: 20))
    }
}
```

If you provide this `.zero()` into the `from` field and sign the transaction later, it will be replaced. But if you really provide the `.zero()` and use `sendTransaction()`, the node will simply fail signing and sending the transaction.    
I think this way it's more obvious that you can't provide `.zero()` if you use `sendTransaction()` and we don't have to throw anything. Nevertheless you will be able to explicitly omit this field if you are going to sign it locally later.

What do you think?

Reply:
I thought about this a bit, and I feel like an optional value is probably more true to the data here. It also makes an error more easy to spot when examining the value, but I know it's not perfect. The goal of this object in my mind is to compose transaction details, which might even happen in different orders (create this object first, then get the current account to fill in the from). It's not technically a validated transaction until it is signed either through sending to the node, or signing locally. Perhaps a name like EthereumTransactionAttributes might make more sense?

The part that makes me lean towards an optional `from` the most is that in web3.js they automatically fill in the `defaultAccount` as the `from` during `sendTransaction` if one is not already provided, so some people are used to leaving that out.

That being said, I would also be okay with implementing this `.zero()` solution if you have a strong preference. If we went that route, is there a way `.zero()` could be called without `try`?

Reply:
I think you are right. The `.zero()` solution is a nasty trick and we shouldn't rely on it. Maybe we should just go for it with the Optional value for now and change it if we have a better solution at some point in time.

Reply:
Do you want to change anything else? If not I am going to merge this...

Reply:
I think it's good, but one thing I thought of is that the readme won't be up to date anymore with the transaction examples. Should that be updated before it's released or after?

Reply:
Please update the README and create a new pull request to the base branch "0.2.0". Once all new features are finished I will merge it into master manually.